base/save.go

```diff
@@ -269,3 +252,75 @@ func GenerateAvailableName(r repo.Repo, peername, prefix string) string {
 		}
 	}
 }
+
+// DatasetNameExists determines whether the name exists in the repository
+// TODO(dustmop): Add dscache support
+func DatasetNameExists(r repo.Repo, dsName string) bool {
+	pro, err := r.Profile()
+	if err != nil {
+		log.Errorf("couldn't get profile: %s", err)
+		return false
+	}
+	peername := pro.Peername
+	lookup := &reporef.DatasetRef{Name: dsName, Peername: peername}
+	err = repo.CanonicalizeDatasetRef(r, lookup)
+	if err == repo.ErrNotFound {
+		return false
+	}
+	return true
+}
+
+// DatasetNameAndStableIdentifers determines the final name for the dataset, by inferring one if
+// necessary, and returns stable identifiers for the full dataset history, and for the most
+// recent version.
+func DatasetNameAndStableIdentifers(ctx context.Context, r repo.Repo, peername string, dsName string, ds *dataset.Dataset, newName bool) (string, string, string, error) {
+	inferredName := MaybeInferName(ds)
+	if inferredName != "" {
+		dsName = inferredName
+	}
+	if DatasetNameExists(r, dsName) {
+		if newName && inferredName != "" {
+			// Using --new flag, name was inferred, but it's already in use. Because the --new
+			// flag was given, user is requesting we invent a unique name. Increment a counter
+			// on the name until we find something that's available.
+			dsName = GenerateAvailableName(r, dsName)
+		} else if newName {
+			// Name was explicitly given, with the --new flag, but the name is already in use.
+			// This is an error.
+			// TODO(dlong): Add a test for this case.
+			return "", "", "", fmt.Errorf("dataset name has a previous version, cannot make new dataset")
```

Comment:
use `github.com/qri-io/qri/errors` to distinguish user feedback from a machine-oriented error. I think we should have an `ErrNameTaken` defined in `base`:
```go
var ErrNameTaken = fmt.Errorf("name already in use")
```
And that should be used here:
```go
return dsref.Ref{}, qerr.New(ErrNameTaken, "dataset name has a previous version, cannot make new dataset")
```

Reply:
Will do. Just to clarify, the parameter on the left is the machine-oriented error, and the parameter on the right is the user-destined message?